### ISSUE TITLE
sympy.stats.sample() API change in 1.7 breaks code and is confusing

### ISSUE THREAD

**Issue by analyzethem** [sympy/sympy#21563]
Spinning off new issue from reemerging discussion on closed issue #19061

`sympy.stats.sample()` has changed behaviour since v1.7, returning an iterator instead of actual sample(s), which is an issue:

- It breaks existing code by breaking the API (now returns an iterator rather than samples)
- The change was AFAIK made without warning about future API change
- The iterator serves no real purpose, as you can only use `next()` on it once
- If you want an iterator (one which works), you can already do that with `sample_iterator()`
- The new behaviour is confusing, non-intuitive and not "pythonic"

Below is a code example that shows the warning message that is triggered, how `sample()` returns an iterator, and how `next()` can only be used once.

```
>>> import sympy.stats as st
>>> X = st.Normal('X', 100, 20)
>>> s = st.sample(X)
C:\Users\analyzethem\Anaconda3\lib\site-packages\sympy\stats\rv.py:1126: UserWarning:
The return type of sample has been changed to return an iterator
object since version 1.7. For more information see
https://github.com/sympy/sympy/issues/19061
  warnings.warn(filldedent(message))
>>> s
<generator object sample_iter.<locals>.return_generator_finite at 0x00000295F020E660>
>>> next(s)
102.97917700376605
>>> next(s)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
StopIteration
```

Personally I discovered this issue as a user, having some code which suddenly broke, and following the link to the issue in the warning. It was a bit difficult figuring out what was going on, help text did not really enlighten me too much, and learning that the iterator can only iterate once was doubly confusing.

If you want a random sample from a stochastic variable `X`, then having to do something like `next(sample(X))` to obtain a single sample is really nonintuitive. Also, if you do e.g. `sample(X, size=100)` and the return value is an iterator, you would really think you can iterate over it to get the 100 values - not that you do `next()` once and that then returns a list of 100 values. This is a bit confusing, and so not the pythonic way.

See #19061 for some comments on the topic. My $0.10 how to deal with it:

- There is a problem to keep the new behaviour (breaks old code for no good reason), or just changing it back (breaks new code)
- Since either option is bad, pick the one that is better long-term (return to the behaviour which makes the most sense, which is the previous one) - but do it from the next release, with a warning
- Put in some interim solution, e.g. adding `sympy.stats.future_sampling()` to change global behaviour for `sample()` to no longer return an iterator (this enables fixing old code by adding one line after the `sympy.stats` import`
- If looking into this piece of the API anyways, it might make sense to also introduce a function `samples()` for requesting more than one sample, so instead of `sample(size=100)` you could do `samples(100)` (cleaner, and without the need for keywords)
- And ... if going "full python", why not add behaviour to stochastic variable objects so that for a stochastic variable `X` you could do `iter(X)` and get a random sample iterator

If there is an issue with a new iterator being created any time `sample()` is called (which seems to be the problem one tried to deal with for #19061), one way would be to deal with that e.g. by having a function `sampler()` which returns an object which is iterable (so from that perspective same behaviour as return value of `sample_iterator()`, but which also has methods for pulling sample(s). So you could do e.g.

```
import sympy.stats as sp
X = st.Normal('X', 100, 20)
s = st.sampler(X)
print('Single sample:', s.sample())
print('Ten samples:', s.samples(10))
print('Ten samples (by iterating):', [next(s) for i in range(10)])
```

By reusing the sampler object for generating samples, you avoid the overhead of creating new sample iterators. You could even deprecate `sample()` for a future major release, in order to use the new `sampler()` feature instead - but `sample()` could continue to exist as a shortcut for `sampler().sample()`

**Comment by oscarbenjamin** [comment]
I think that the primary goal should be to restore something that is compatible with the old behaviour at least in common cases but that doesn't have the inconsistency problems of the old API. That means focussing on the likely most common usage of `sample`.

If the behaviour is changed in the next release then why would there be a warning? That seems to repeat the mistake of adding a warning after the behaviour is changed. I don't see that as a good use of warnings - if the behaviour breaks something then the user likely sees an error anyway. Warnings should be used when we want the user to change their code but we don't want to break their code just yet.

The `future_sampling` suggestion would mean adding a function that was only available in the new releases. Since it wouldn't be available in previous releases it doesn't really help a library author who is trying to straddle multiple sympy releases. If they have to write their code for the newest version of sympy then it doesn't help much either because they can just use the new changed behaviour.

**Comment by mostlyaman** [comment]
Hi, while trying to fix the `sympy.stats.sample()` function, I have been able to make it return values instead of an iterator.
```python
>>>from sympy.stats import *
>>> X = Normal("X", 100, 10)
>>> sample(X)
98.92116454643755
>>> sample(X, size = 5)
array([113.5067288 ,  98.23302408, 126.06587997, 100.38151336,
        90.69890246])
```
I wish to start over with the problem and try to implement `sampler()` function concept as described. I like the idea of integrating `sample()`, `samples(N)` and `next(iterable_obj)` within a single `sampler` class.

I made a temporary class to show what it would look like. I hope I am going in the right direction.
```python
>>> X = Normal("X", 100, 20)
>>> Y = sampler(X)
>>> type(Y)
<class '__main__.sampler'>
>>> Y.sample()

94.97939081554124
>>> 
>>> Y.samples(4)
array([103.62854547, 118.52493789,  91.56912261, 102.5029822 ])
>>> next(Y)
84.22851115656765
>>> for i in range(10):
...     next(Y)
... 
89.6284971250255
89.02257810356126
65.42008098180727
94.85793563389163
135.88541914692917
113.2483861781903
83.96442553012953
131.52220607556401
100.18216789896603
90.69935601367081
```

**Comment by oscarbenjamin** [comment]
@czgdp1807 @Upabjojr what do you think?

**Comment by czgdp1807** [comment]
> I like the idea of integrating sample(), samples(N) and next(iterable_obj) within a single sampler class.

+1. May be everything encapsulated under, `class Sampler` would be great IMHO.

**Comment by mostlyaman** [comment]
Regarding the future of `sample` and `sample_iter`, do they need to be removed or can they co-exist with the new `sampler` class.
If they can co-exist, then we can use them and `sampler` class would just be more user friendly interface to get an output. It also reduces the chances of breaking already written code.

**Comment by oscarbenjamin** [comment]
We shouldn't remove `sample` and `sample_iter`. This issue starts with the problem that a backwards incompatible change was made so we shouldn't make another incompatible change.

**Comment by mostlyaman** [comment]
So, we should aim to accommodate changes that re-enable the Behaviour before 1.7 as well as fix the inconsistency of sample returning an iterator after 1.7

**Comment by oscarbenjamin** [comment]
Yes, I think so. We should fix `sample` so that previous usage works again at least in basic cases.

Adding new API like `sampler` is a separate question although that could be done at the same time if the design is agreed.

@Upabjojr what do you think about sampler?

**Comment by mostlyaman** [comment]
`sample` before v1.7 only returned single values as outputs. I hope this matches its behavior. However, I kept the size parameter which was not present before v1.7 as it may be helpful to some users.
```python
>>> from sympy.stats import Normal, sample
>>> Y = Normal('Y', 10, 2)
>>> sample(Y)
11.305896549354351
>>> sample(Y, size=3)
array([8.49990134, 8.55926544, 9.29693864])

```

**Comment by oscarbenjamin** [comment]
The size argument was present in sympy 1.6 and returned a list.

**Comment by mostlyaman** [comment]
Oh, thanks for the correction! I checked on sympy-live for an older version of sympy and that was running v1.5
Do we need any other changes for this issue?

**Comment by oscarbenjamin** [comment]
I think that we just need to restore sensible behaviour for `sample` that is at least partly compatible with sympy 1.6. With 1.6 we have:
```python
In [1]: from sympy.stats import sample, Uniform

In [2]: U = Uniform('U', 0, 1)

In [3]: sample(U, size=3)
Out[3]: [0.705006858786766, 0.705006858786766, 0.705006858786766]

In [4]: sample(U)
Out[4]: 0.834900756535122
```
In sympy 1.8 there is a size argument and also a numsamples argument that do this:
```python
In [11]: list(sample(U))
Out[11]: [0.7687745595834666]

In [12]: list(sample(U, size=3))
Out[12]: [array([0.81036993, 0.18580399, 0.01911709])]

In [13]: list(sample(U, numsamples=4))
Out[13]: [0.8443370958718625, 0.5233633248430264, 0.44949831514077154, 0.6642675310525122]

In [14]: list(sample(U, size=3, numsamples=4))
Out[14]: 
[array([0.08571217, 0.12133931, 0.55081785]),
 array([0.91329635, 0.59790493, 0.06630375]),
 array([0.98378562, 0.3421354 , 0.25134057]),
 array([0.0780444 , 0.553857  , 0.68438741])]

In [15]: list(sample(U, size=(3, 2), numsamples=4))
Out[15]: 
[array([[0.1409835 , 0.59641373],
        [0.66439721, 0.43545229],
        [0.69913758, 0.15875101]]),
 array([[0.65461532, 0.99656592],
        [0.95069699, 0.62747994],
        [0.84766795, 0.25896363]]),
 array([[0.33467659, 0.2933884 ],
        [0.21333459, 0.13437935],
        [0.80524673, 0.88757811]]),
 array([[0.33923549, 0.10576515],
        [0.23662182, 0.85352809],
        [0.86694497, 0.94047964]])]
```

**Comment by mostlyaman** [comment]
**OUTPUTS**
I think I have been able to produce something sensible for `sample`.
```python
>>> from sympy.stats import *
>>> X = Normal('X', 10, 2)
>>> sample(X)
7.759764913600893

>>> sample(X, size=3)
[10.242212269115184, 10.910001528435632, 5.843371943884264]

>>> sample(X, numsamples=2)
[9.225171331294497, 8.279742564372047]

>>> sample(X, size=2, numsamples=3)
[array([10.85110748,  7.87308432]),
 array([7.5702038 , 5.55675533]),
 array([ 7.39925658, 12.66477779])]

>>> sample(X, size=(3,2))
array([[10.22989246, 12.66969921],
       [10.00457022, 10.70168089],
       [12.33775328,  7.81722807]])

>>> sample(X, size=(3,2), numsamples=4)
[array([[13.34077277,  7.98441588],
       [11.17882582, 12.46614747],
       [ 9.44281301,  9.26682168]]), 
array([[ 7.1304523 ,  9.02925785],
       [12.67305056,  7.79069383],
       [12.2258876 ,  7.77003153]]),
array([[ 6.43758548, 11.03560918],
       [10.04824432,  7.94649574],
       [13.10339154,  6.43410866]]), 
array([[10.73569628, 11.46022168],
       [ 7.21191141,  9.92608357],
       [12.04964084, 10.54806433]])]
```
**PROBLEMS**
Only problem I am having is with this example.
>In [11]: list(sample(U))
>Out[11]: [0.7687745595834666]

Since, sample(U) returns a float now, it causes an error.
```python
>>> list(sample(X))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'numpy.float64' object is not iterable
```
**CHANGES**
The changes I have done are as follows.
```python
def sample(expr, condition=None, size=(), library='scipy', numsamples=1, seed=None, **kwargs):
    iterator = sample_iter(expr, condition, size=size, library=library, numsamples=numsamples, seed=seed)
    #case for numsamples being used
    if numsamples != 1:
    	return [next(iterator) for i in range(numsamples)]
    #case for size=integer
    if type(size) == type(0):
    	return list(next(iterator))
    #case for size=tuple
    if type(size) == type(()):
    	return next(iterator)
```

**Comment by oscarbenjamin** [comment]
> Since, sample(U) returns a float now, it causes an error.

I think that's fine. It's more important to be consistent with the version of `sample` from before it was changed to return an iterator.

**Comment by mostlyaman** [comment]
So, if the change is okay, should I make the PR? The documentation also needs to be updated.

**Comment by oscarbenjamin** [comment]
The basic issue has been fixed by #21590 which is now merged to master.

@analyzethem  can you test the current sympy master?
